lib/AST/ASTContext.cpp

```diff
@@ -3247,17 +3247,7 @@ PackExpansionType::PackExpansionType(Type patternType, Type countType,
                                      RecursiveTypeProperties properties,
                                      const ASTContext *canCtx)
   : TypeBase(TypeKind::PackExpansion, canCtx, properties),
-    patternType(patternType), countType(countType) {
-
-  // TODO: it would be nice if the solver didn't stick PlaceholderTypes and
-  // UnresolvedTypes in here.
-  assert(countType->is<PackType>() ||
```

Comment:
My original intent here was that the caller must either provide a valid count type or not build a PackExpansionType at all (and instead return, eg, ErrorType). Can you explain the scenario where this plan fails?

Reply:
I know, but the solver builds invalid pack expansion types through substitution when a shape type is invalid, e.g. because the pack has no `each` expressions inside. In that case `getReducedShape` will return `Void` when simplifying the `ShapeOf` constraint, and the solver will create `PackExpansionType(patternType: Int, shapeType: Void)`. I don't think making the solver replace `PackExpansionType`s with `ErrorType`s is any better than what I've done here, especially because precise information in invalid solutions is very helpful for diagnostics. I can try instead to bind a hole type via the `ShapeOf` constraint, but I wouldn't be surprised if we rely on successfully solving those constraints for concrete types in valid solutions.

It's also possible we can come up with a better representation for invalid pack expansion types, but until then, this patch is so much more tolerant of invalid code.